packages/next/src/server/app-render/cache-signal.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * This class is used to detect when all cache reads for a given render are settled.
+ * We do this to allow for cache warming the prerender without having to continue rendering
+ * the remainder of the page. This feature is really only useful when the dynamicIO flag is on
+ * and should only be used in codepaths gated with this feature.
+ */
+
+export class CacheSignal {
+  private count: number
+  private listeners: Array<() => void>
+
+  constructor() {
+    this.count = 0
+    this.listeners = []
+  }
+
+  /**
+   * If there are inflight cache reads this Promise can resolve in a microtask however
+   * if there are no inflight cache reads then we wait at least one task to allow initial
+   * cache reads to be initiated.
+   */
+  async cacheReady() {
+    const promise = new Promise<void>((resolve) => {
```

Comment:
Better to use a `packages/next/src/lib/detached-promise.ts` here?

Reply:
generally prefer to use platform primitives if I can. I don't think it would meanignfully clarify the code